File: src/sessionManager.ts
import { config } from "./config";
import { LinkedInAuthError, LinkedInBlockedError, SessionFlaggedError } from "./errors";
import { fetchProfileView } from "./linkedinClient";
import { getSessionHealth, setSessionHealth } from "./store/sessionHealthStore";

export { getSessionHealth };

// Wraps the raw LinkedIn client with two protections most naive scrapers skip:
//
// 1. Rate limiting with jitter — paces requests like a human browsing, not a script firing as
//    fast as jobs arrive. This is enforced here (not per-request in the API layer) so it holds
//    across the whole worker regardless of how many jobs pile up.
// 2. A circuit breaker — after repeated auth/block failures, stop hitting LinkedIn entirely
//    instead of hammering an already-flagged or dead session into a permanent ban.
let nextAvailableAt = 0;

async function throttle(): Promise<void> {
  const now = Date.now();
  const waitMs = Math.max(0, nextAvailableAt - now);
  if (waitMs > 0) {
    await new Promise((resolve) => setTimeout(resolve, waitMs));
  }
  const jitter = Math.random() * config.requestJitterMs;
  nextAvailableAt = Date.now() + config.minRequestIntervalMs + jitter;
}

export async function fetchProfileThroughSession(publicIdentifier: string): Promise<unknown> {
  const health = getSessionHealth();
  if (health.state === "flagged") {
    throw new SessionFlaggedError(health.lastError);
  }

  await throttle();

  try {
    const raw = await fetchProfileView(publicIdentifier);
    setSessionHealth({
      ...health,
      state: "healthy",
      consecutiveFailures: 0,
      lastRequestAt: new Date().toISOString(),
    });
    return raw;
  } catch (err) {
    if (err instanceof LinkedInAuthError || err instanceof LinkedInBlockedError) {
      const consecutiveFailures = health.consecutiveFailures + 1;
      const flagging = consecutiveFailures >= config.sessionFailureThreshold;
      const now = new Date().toISOString();
      setSessionHealth({
        state: flagging ? "flagged" : "healthy",
        consecutiveFailures,
        lastError: err.message,
        lastErrorAt: now,
        flaggedAt: flagging ? now : health.flaggedAt,
        lastRequestAt: now,
      });
    }
    throw err;
  }
}

export function resetSessionHealth(): void {
  setSessionHealth({
    state: "healthy",
    consecutiveFailures: 0,
    lastError: null,
    lastErrorAt: null,
    flaggedAt: null,
    lastRequestAt: null,
  });
}

File: app/session_manager.py
import asyncio
import random
import time
from datetime import datetime, timezone
from typing import Any

from . import stores
from .config import config
from .errors import LinkedInAuthError, LinkedInBlockedError, SessionFlaggedError
from .linkedin_client import fetch_profile_html

# Wraps the raw client with the two protections a naive scraper skips:
#
#   1. Pacing with jitter — enforced here rather than per-request so it holds across the whole
#      worker no matter how many jobs queue up.
#   2. A circuit breaker — after repeated auth/block failures we stop entirely instead of
#      pushing an already-flagged session toward a hard account restriction.
#
# Both matter more than they look: an invalidated session logs the account out everywhere and
# can only be recovered by a manual browser login.
_next_available_at = 0.0
_lock = asyncio.Lock()


async def _throttle() -> None:
    global _next_available_at
    async with _lock:
        wait = _next_available_at - time.monotonic()
        if wait > 0:
            await asyncio.sleep(wait)
        _next_available_at = (
            time.monotonic()
            + config.min_request_interval_s
            + random.random() * config.request_jitter_s
        )


async def fetch_profile_through_session(public_identifier: str) -> Any:
    health = stores.get_session_health()
    if health["state"] == "flagged":
        raise SessionFlaggedError(health.get("lastError"))

    await _throttle()

    try:
        raw = await fetch_profile_html(public_identifier)
    except (LinkedInAuthError, LinkedInBlockedError) as err:
        failures = health["consecutiveFailures"] + 1
        flagging = failures >= config.session_failure_threshold
        now = datetime.now(timezone.utc).isoformat()
        stores.set_session_health(
            {
                "state": "flagged" if flagging else "healthy",
                "consecutiveFailures": failures,
                "lastError": str(err),
                "lastErrorAt": now,
                "flaggedAt": now if flagging else health.get("flaggedAt"),
                "lastRequestAt": now,
            }
        )
        raise

    stores.set_session_health(
        {
            **health,
            "state": "healthy",
            "consecutiveFailures": 0,
            "lastRequestAt": datetime.now(timezone.utc).isoformat(),
        }
    )
    return raw
